packages/autocertifier-client/src/AutoCertifierClient.ts

```diff
@@ -0,0 +1,172 @@
+import { EventEmitter } from 'eventemitter3'
+import { IAutoCertifierService } from './proto/packages/autocertifier/protos/AutoCertifier.server'
+import { SessionIdRequest, SessionIdResponse } from './proto/packages/autocertifier/protos/AutoCertifier'
+import { ServerCallContext } from '@protobuf-ts/runtime-rpc'
+import { filePathToNodeFormat } from '@streamr/utils'
+import { RestClient } from './RestClient'
+import { CertifiedSubdomain } from './data/CertifiedSubdomain'
+import fs from 'fs'
+import path from 'path'
+import * as forge from 'node-forge'
+import { Logger } from '@streamr/utils'
+
+interface AutoCertifierClientEvents {
+    updatedSubdomain: (domain: CertifiedSubdomain) => void
+}
+
+const logger = new Logger(module)
+
+export const AUTOCERTIFIER_SERVICE_ID = 'system/auto-certificer'
+const ONE_DAY = 1000 * 60 * 60 * 24
+const MAX_INT_32 = 2147483647
+
+export class AutoCertifierClient extends EventEmitter<AutoCertifierClientEvents> implements IAutoCertifierService {
+
+    private updateTimeout?: NodeJS.Timeout
+    private readonly restClient: RestClient
+    private readonly subdomainPath: string
+    private readonly streamrWebSocketPort: number
+    private readonly ongoingSessions: Set<string> = new Set()
+
+    constructor(
+        subdomainPath: string,
+        streamrWebSocketPort: number,
+        restApiUrl: string,
+        registerRpcMethod: (
+            serviceId: string,
+            rpcMethodName: string,
+            method: (request: SessionIdRequest, context: ServerCallContext) => Promise<SessionIdResponse>
+        ) => void
+    ) {
+        super()
+
+        this.restClient = new RestClient(restApiUrl)
+        this.subdomainPath = filePathToNodeFormat(subdomainPath)
+        this.streamrWebSocketPort = streamrWebSocketPort
+        registerRpcMethod(AUTOCERTIFIER_SERVICE_ID, 'getSessionId', this.getSessionId.bind(this))
+    }
+
+    public async start(): Promise<void> {
+        if (!fs.existsSync(this.subdomainPath)) {
+            await this.createCertificate()
+        } else {
+            await this.checkSubdomainValidity()
+        }
+    }
+
+    private async checkSubdomainValidity(): Promise<void> {
+        const sub = this.loadSubdomainFromDisk()
+
+        if (Date.now() >= sub.expirationTimestamp - ONE_DAY) {
+            await this.updateCertificate()
+        } else {
+            await this.updateSubdomainIpAndPort()
+            this.scheduleCertificateUpdate(sub.expirationTimestamp)
+            this.emit('updatedSubdomain', sub.subdomain)
+        }
+    }
+
+    private loadSubdomainFromDisk(): { subdomain: CertifiedSubdomain, expirationTimestamp: number } {
+        const subdomain = JSON.parse(fs.readFileSync(this.subdomainPath, 'utf8')) as CertifiedSubdomain
+        const certObj = forge.pki.certificateFromPem(subdomain.certificate.cert)
+        const expirationTimestamp = certObj.validity.notAfter.getTime()
+        return { subdomain, expirationTimestamp }
```

Comment:
Could just return `subdomain`, and the expiration timestamp parsing could be done in `checkSubdomainValidity`